File: implementations/typescript/packages/recovery/test/delivery-journal.test.js
import assert from "node:assert/strict";
import test from "node:test";
import { DeliveryJournal } from "@axisrobo/harmovela-recovery";

test("DeliveryJournal appends events with sequence", () => {
  const journal = new DeliveryJournal();
  const seq1 = journal.append({ type: "task.submitted", task_id: "task_01" });
  const seq2 = journal.append({ type: "task.completed", task_id: "task_01" });
  assert.equal(seq1, 1);
  assert.equal(seq2, 2);
});

test("DeliveryJournal replays events since cursor", () => {
  const journal = new DeliveryJournal();
  journal.append({ type: "task.submitted" });
  journal.append({ type: "task.started" });
  journal.append({ type: "task.completed" });

  const events = journal.replay("stream_01:1");
  assert.equal(events.length, 2);
  assert.equal(events[0].type, "task.started");
  assert.equal(events[1].type, "task.completed");
});

test("DeliveryJournal replays all events with default cursor", () => {
  const journal = new DeliveryJournal();
  journal.append({ type: "task.submitted" });
  journal.append({ type: "task.started" });

  const events = journal.replay();
  assert.equal(events.length, 2);
});

test("DeliveryJournal purges events before cursor", () => {
  const journal = new DeliveryJournal();
  journal.append({ type: "evt_1" });
  journal.append({ type: "evt_2" });
  journal.append({ type: "evt_3" });

  const removed = journal.purge("stream_01:2");
  assert.equal(removed, 2);

  const events = journal.replay();
  assert.equal(events.length, 1);
  assert.equal(events[0].type, "evt_3");
});

test("DeliveryJournal provides stats", () => {
  const journal = new DeliveryJournal();
  journal.append({ type: "evt_1" });
  journal.append({ type: "evt_2" });

  const stats = journal.getStats();
  assert.equal(stats.totalEvents, 2);
  assert.equal(stats.oldestSequence, 1);
  assert.equal(stats.newestSequence, 2);
});

test("DeliveryJournal stats are empty for new journal", () => {
  const journal = new DeliveryJournal();
  const stats = journal.getStats();
  assert.equal(stats.totalEvents, 0);
  assert.equal(stats.oldestSequence, null);
  assert.equal(stats.newestSequence, null);
});

File: implementations/java/src/test/java/com/axisrobo/harmovela/recovery/DeliveryJournalTest.java
package com.axisrobo.harmovela.recovery;

import org.junit.jupiter.api.Test;
import java.util.Map;
import static org.junit.jupiter.api.Assertions.*;

class DeliveryJournalTest {

    @Test
    void appendsEventsWithSequence() {
        var journal = new DeliveryJournal();
        var seq1 = journal.append(Map.of("type", "task.submitted", "task_id", "task_01"));
        var seq2 = journal.append(Map.of("type", "task.completed", "task_id", "task_01"));
        assertEquals(1, seq1);
        assertEquals(2, seq2);
    }

    @Test
    void replaysEventsSinceCursor() {
        var journal = new DeliveryJournal();
        journal.append(Map.of("type", "task.submitted"));
        journal.append(Map.of("type", "task.started"));
        journal.append(Map.of("type", "task.completed"));
        var events = journal.replay("stream_01:1");
        assertEquals(2, events.size());
        assertEquals("task.started", events.get(0).get("type"));
        assertEquals("task.completed", events.get(1).get("type"));
    }

    @Test
    void replaysAllEventsWithDefaultCursor() {
        var journal = new DeliveryJournal();
        journal.append(Map.of("type", "task.submitted"));
        journal.append(Map.of("type", "task.started"));
        var events = journal.replay(null);
        assertEquals(2, events.size());
    }

    @Test
    void purgesEventsBeforeCursor() {
        var journal = new DeliveryJournal();
        journal.append(Map.of("type", "evt_1"));
        journal.append(Map.of("type", "evt_2"));
        journal.append(Map.of("type", "evt_3"));
        var removed = journal.purge("stream_01:2");
        assertEquals(2, removed);
        var events = journal.replay(null);
        assertEquals(1, events.size());
        assertEquals("evt_3", events.get(0).get("type"));
    }

    @Test
    void providesStats() {
        var journal = new DeliveryJournal();
        journal.append(Map.of("type", "evt_1"));
        journal.append(Map.of("type", "evt_2"));
        var stats = journal.getStats();
        assertEquals(2, stats.get("totalEvents"));
        assertEquals(1, stats.get("oldestSequence"));
        assertEquals(2, stats.get("newestSequence"));
    }

    @Test
    void statsAreEmptyForNewJournal() {
        var journal = new DeliveryJournal();
        var stats = journal.getStats();
        assertEquals(0, stats.get("totalEvents"));
        assertNull(stats.get("oldestSequence"));
        assertNull(stats.get("newestSequence"));
    }
}
